src/main/java/com/google/api/gax/grpc/OperationFuture.java

```
@@ -0,0 +1,372 @@
+        Class<ResponseT> responseClass,
+        Duration pollingInterval,
+        Waiter waiter,
+        CountDownLatch asyncCompletionLatch) {
+      this.startOperationFuture = startOperationFuture;
+      this.finalResultFuture = finalResultFuture;
+      this.operationsApi = operationsApi;
+      this.responseClass = responseClass;
+      this.pollingInterval = pollingInterval;
+      this.waiter = waiter;
+      this.asyncCompletionLatch = asyncCompletionLatch;
+    }
+
+    @Override
+    public ResponseT call() {
+      try {
+        callImpl();
+      } finally {
+        asyncCompletionLatch.countDown();
+      }
+      return null;
+    }
+
+    public void callImpl() {
+      Operation firstOperation = null;
+      try {
+        firstOperation = startOperationFuture.get();
+        if (firstOperation.getDone()) {
+          setResultFromOperation(finalResultFuture, firstOperation, responseClass);
+          return;
+        }
+        do {
+          // TODO: switch implementation from polling to scheduled execution
+          waiter.wait(pollingInterval);
+          Operation latestOperation = operationsApi.getOperation(firstOperation.getName());
+          if (latestOperation.getDone()) {
+            if (isCancelled(latestOperation)) {
+              finalResultFuture.cancel(true);
+            } else {
+              setResultFromOperation(finalResultFuture, latestOperation, responseClass);
+            }
+            return;
+          }
+        } while (true);
+      } catch (InterruptedException e) {
+        try {
+          if (firstOperation != null) {
+            operationsApi.cancelOperation(firstOperation.getName());
+          }
+          if (!startOperationFuture.isDone()) {
+            startOperationFuture.cancel(true);
+          }
+          finalResultFuture.cancel(true);
+
+          Thread.currentThread().interrupt();
+          return;
+        } catch (Exception e2) {
+          finalResultFuture.cancel(true);
```

Are we ignoring the exception here? Shouldn't we call `finalResultFuture.setException(e2);`?